From: https://patch-diff.githubusercontent.com/raw/MariaDB/server/pull/2896.patch
Author: Tuukka Pasanen <tuukka.pasanen@ilmi.fi>
Date: Mon, 4 Dec 2023 10:24:07 +0200
Subject: [PATCH] MDEV-32936: Detect do ARM processor have NEON intrinsics or
 not

NEON was introcuded as optional when ARMv7 was launched.
ARMv8 should have NEON buildin but there is still ARMv8-R
specification which does have optional NEON technology.

As said above Neon is not available lower than ARmv7 architecture CPUs
which are still widely available inside NAS devices.

PLEASE NOTE: ARMv5, ARMv6 or ARMv7 architectures are not currently
offiacially supported by upstream.
---
 mysys/CMakeLists.txt | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

--- a/mysys/CMakeLists.txt
+++ b/mysys/CMakeLists.txt
@@ -88,6 +88,11 @@ ELSEIF(CMAKE_SYSTEM_PROCESSOR MATCHES "a
   IF(CMAKE_COMPILER_IS_GNUCC)
     include(CheckCXXSourceCompiles)
 
+    # Check if NEON intrinsics are available and does not
+    # print error 'NEON intrinsics not available with the soft-float ABI.
+    # Please use -mfloat-abi=softfp or -mfloat-abi=hard'
+    CHECK_INCLUDE_FILES(arm_neon.h HAVE_ARMV8_NEON)
+
     CHECK_CXX_SOURCE_COMPILES("
     #define CRC32CX(crc, value) __asm__(\"crc32cx %w[c], %w[c], %x[v]\":[c]\"+r\"(crc):[v]\"r\"(value))
     asm(\".arch_extension crc\");
@@ -134,9 +139,21 @@ ELSEIF(CMAKE_SYSTEM_PROCESSOR MATCHES "a
       IF(HAVE_ARMV8_CRYPTO)
         ADD_DEFINITIONS(-DHAVE_ARMV8_CRYPTO)
       ENDIF()
+      IF(NOT HAVE_ARMV8_NEON)
+        set(CMAKE_REQUIRED_FLAGS "-mfpu=neon -mfloat-abi=softfp")
+        CHECK_INCLUDE_FILES(arm_neon.h HAVE_ARMV8_NEON_SOFTFP)
+        set(CMAKE_REQUIRED_FLAGS "")
+      ENDIF()
       SET(MYSYS_SOURCES ${MYSYS_SOURCES} crc32/crc32_arm64.c)
+
+      IF(HAVE_ARMV8_NEON_SOFTFP)
+        SET(CRC32_ARM64_FLAGS "-mfpu=neon -mfloat-abi=softfp -march=armv8-a+crc+crypto")
+      ELSE()
+        SET(CRC32_ARM64_FLAGS "-march=armv8-a+crc+crypto")
+      ENDIF()
+
       SET_SOURCE_FILES_PROPERTIES(crc32/crc32_arm64.c PROPERTIES
-        COMPILE_FLAGS "-march=armv8-a+crc+crypto")
+        COMPILE_FLAGS ${CRC32_ARM64_FLAGS})
     ENDIF()
   ENDIF()
 ENDIF()
